infra/src/main/kotlin/fi/fta/geoviite/api/frameconverter/v1/FrameConverterControllerV1.kt

```diff
@@ -1,115 +1,89 @@
 import fi.fta.geoviite.infra.common.LayoutContext
 import fi.fta.geoviite.infra.common.MainLayoutContext
-import fi.fta.geoviite.infra.error.IntegrationApiExceptionV1
-import fi.fta.geoviite.infra.localization.LocalizationLanguage
-import fi.fta.geoviite.infra.localization.LocalizationService
+import fi.fta.geoviite.infra.error.ExtApiExceptionV1
 import org.springframework.beans.factory.annotation.Autowired
-import org.springframework.context.annotation.Bean
-import org.springframework.context.annotation.Configuration
-import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder
 import org.springframework.security.access.prepost.PreAuthorize
-import org.springframework.web.bind.annotation.RequestMapping
-import org.springframework.web.bind.annotation.RequestMethod
+import org.springframework.web.bind.annotation.GetMapping
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.RequestBody
 import org.springframework.web.bind.annotation.RequestParam
 
-@Configuration
-class FrameConverterApiObjectMapperV1 {
-
-    @Bean
-    fun objectMapper(builder: Jackson2ObjectMapperBuilder): ObjectMapper {
-        val module = SimpleModule()
-        module.addDeserializer(FrameConverterRequestV1::class.java, FrameConverterRequestDeserializerV1())
-        return builder.build<ObjectMapper>().registerModule(module)
-    }
-}
-
 @PreAuthorize(AUTH_API_FRAME_CONVERTER)
-@GeoviiteIntegrationApiController(
-    [
-        "/rata-vkm",
-        "/rata-vkm/dev",
-        "/rata-vkm/v1",
-        "/rata-vkm/dev/v1",
-
-        // Trailing slashes are also supported in the frame converter.
-        "/rata-vkm/",
-        "/rata-vkm/dev/",
-        "/rata-vkm/v1/",
-        "/rata-vkm/dev/v1/",
-    ]
-)
+@GeoviiteExtApiController(["/rata-vkm/v1", "/rata-vkm/dev/v1"])
 class FrameConverterControllerV1
 @Autowired
-constructor(
-    private val objectMapper: ObjectMapper,
-    private val frameConverterServiceV1: FrameConverterServiceV1,
-    private val localizationService: LocalizationService,
-) {
+constructor(private val objectMapper: ObjectMapper, private val frameConverterServiceV1: FrameConverterServiceV1) {
 
-    @RequestMapping(method = [RequestMethod.GET, RequestMethod.POST], params = ["json"])
-    fun multiInputTransform(
-        @RequestParam(name = "json") requests: List<FrameConverterRequestV1>?
-    ): GeoJsonFeatureCollection {
-        val parsedRequests = requests ?: emptyList()
+    @GetMapping("/koordinaatit", "/koordinaatit/")
+    fun trackAddressToCoordinateRequestSingle(@RequestParam params: Map<String, String?>): GeoJsonFeatureCollection {
+        val jsonString = objectMapper.writeValueAsString(params)
+        val request = objectMapper.readValue(jsonString, TrackAddressToCoordinateRequestV1::class.java)
+        val queryParams = objectMapper.readValue(jsonString, FrameConverterQueryParamsV1::class.java)
```

Comment:
Probleema: Miten mäpätä yksinkertaisesti suomenkieliset queryParamit englanninkielisiin dataclass-kenttiin? Ei onnistu ollenkaan yhtä yksiselitteisesti kuin jsonin ja jacksonin kanssa. Nykyinen ratkaisu siis on hieman ankea, se muuttaa queryparamit json-stringiksi ja parsii ne sieltä.

* Lisää overheadiä muodostaa json ja parsia se uudelleen. Samalla sarjassamme "tuntuu typerältä edes tehdä näin"
* GETtiä ja POSTa pitää kumminkin tukea, joten jsonia varten tarvitaan Jackson-tuki kuitenkin
*  Suomi-englanti -kenttämäppäystä olisi mielekästä ylläpitää vain yhdessä paikassa saada eikä useammassa (tällä rakenteella siis pelkästään @JsonProperty-annotaatiolla samassa dataluokassa)
* Custom-annotaatio joka tulee kumpaakin (QueryParamia ja JsonPropertyä samanaikaisesti) on aika njärsyttävä aikaansaada (en selvitellyt loppuun asti onko tämä edes mahdollista, mutta edellyttää vähintään melkoisia kommervenkkejä)
* Jokainen nykyisistä Controllereista käsittelee myös queryParameita (myös POST-controllerit), joten jos näissä olisi jokaisessa @QueryParam jokaiselle queryParamille, ettei väliaikaisjsonia tarvitsisi muodostaa (vaan controlleri muodostaisi bindattujen muuttujien perusteella dataluokan), niin siitä syntyisi aika reippaasti toisteisuutta controllereihin.

Lopputulos: Serialisoidaan queryParamit jsoniksi ja hyväksytään typeryys tehdä näin, jolloin käännösmäppäystä tarvitsee tehdä vain yhdessä paikassa, hämäriä custom-annotaatioita ei tarvitse tehdä tai ylläpitää, koodi pysyy helppolukuisena sekä lyhyenä ilman toisteisuutta. JSON-overhead (joka toki varmaan pientä verrattuna varsinaiseen laskuoperaatioon) on sinänsä myös ihan siedettävä kompromissi tässä tilanteessa.

Parempia idiksiä?

Reply:
Voi olla että en ihan täysin tajunnut ongelmaa, sillä tuo olisi mielestäni hoidettavissa niillä mekanismeilla joita FrameConverter käyttää jo siinä mainin versiossa joka minulla on tällä hetkellä. `@RequestParam`:han mahdollistaa kutsussa tulevien parametrien "uudelleennimeämisen" muuttujiksi (`@RequestParam("parametriSuomeksi") paramInEnglish`) ja dataluokat tukevat vastaavaa `@JsonProperty`-annotaatiolla (`@JsonProperty("kenttaSuomeksi") fieldInEnglish`)

Reply:
Siis `@RequestParam("parametriSuomeksi")` edellyttää jokaisen parametrin toistamista jokaiseen controlleriin, ja parametridataluokan rakentamisen. Lisäksi silloin tuo mäppäys olisi kahdessa paikassa eli `@RequestParam`-annotaatiossa ja erikseen `@JsonPropertynä` varsinaisille pyynnöille

Lähinnä siis toisteisuutta yritetään tuolla välttää + helpottaa ylläpitoa

Reply:
Aivan joo. En itse oikeastaan nää tuota parametrien toistoa niin huonona asiana, vaikka niitä onkin toki ärsyttävää kirjoittaa. Se selkeyttää lukijalle todella paljon sitä mitä kontrolleri hyväksyy sisäänsä ja antaa eteenpäin, sillä esimerkiksi kenttien pakollisuuteen/optionaalisuuteen tulee ottaa kantaa jo tuossa vaiheessa ja helpottaa oikean controllerin löytämistä findilla parametrien perusteella

Reply:
Juu, on siinä puolensa, onhan tämä vähän erikoinen rakenne. Jutskaillaan dailyn yhteydessä vaikkapa

Jos kaikilla controlleilla on vähintään samat queryParamit (3x asetus+1x koordinaattiparametrisointi), joista kaikki optionaalisia (default-arvojen vuoksi), samoilla nimillä, niin ei se kyllä hakua varmaan juuri helpota... 🤔

Reply:
Tuli myös yksi lisäpointti mieleen: Jos esim. uuden "rata-vkm"-globaalin paluuarvoasetuksen tai muun parametrisoinnin haluaisi lisätä, niin tällä queryparamit-jsoniksi rakenteella riittää lisätä yksi kenttä FrameConverterQueryParams-dataluokkaan. Jos nuo olisi kaikkialla erikseen, niin sen uuden myös joutuisi lisäämään jokaiseen paikkaan vähintään kahdesti (queryparam-bindaus ja parametridataluokan rakentajakutsu). Tämä olisi myös entistä raskaampaa mitä enemmän noita controllereita olisi (ei kovin skaalautuva rakenne) 🤔

Reply:
Aika kumsi-kumsa. Hiukan olen kyllä Juusen linjoilla että toistaminen on ehkä pienempi synti kuin json luonti/deserialisointi erikoisuus. Jos nimen toisteisuutta haluaa välttää niin kai sen voisi vakioida, mutta en oikeastaan näe sitäkään tarpeelliseksi. Toisaalta argumentteja lisätessä olisi ihan luonnollista että näissä yksittäisen endpointeissa voidaan erikseen päättää otetaanko samat argumentit, mutta koska ne menee eteenmpäin kuitenkin samana niin... noh. Molempi huonompi.